dask/dataframe/io/parquet/core.py

```diff
@@ -701,60 +759,56 @@ def to_parquet(
         **kwargs_pass,
     )
 
-    # check name_function is valid
-    if name_function is not None and not callable(name_function):
-        raise ValueError("``name_function`` must be a callable with one argument.")
-
-    # Use i_offset and df.npartitions to define file-name list
-    filenames = [
-        f"part.{i + i_offset}.parquet"
-        if name_function is None
-        else name_function(i + i_offset)
-        for i in range(df.npartitions)
-    ]
-
-    if name_function is not None and len(set(filenames)) < len(filenames):
-        raise ValueError("``name_function`` must produce unique filenames.")
+    # Check that custom name_function is valid,
+    # and that it will produce unique names
+    if name_function is not None:
+        if not callable(name_function):
+            raise ValueError("``name_function`` must be a callable with one argument.")
+        filenames = [name_function(i + i_offset) for i in range(df.npartitions)]
+        if len(set(filenames)) < len(filenames):
+            raise ValueError("``name_function`` must produce unique filenames.")
 
-    # Construct IO graph
-    dsk = {}
-    name = "to-parquet-" + tokenize(
-        df,
-        fs,
-        path,
-        append,
-        ignore_divisions,
-        partition_on,
-        division_info,
-        index_cols,
-        schema,
-    )
-    part_tasks = []
+    # Create Blockwise layer for parquet-data write
     kwargs_pass["fmd"] = meta
     kwargs_pass["compression"] = compression
     kwargs_pass["index_cols"] = index_cols
     kwargs_pass["schema"] = schema
-    for d, filename in enumerate(filenames):
-        dsk[(name, d)] = (
-            apply,
-            engine.write_partition,
-            [
-                (df._name, d),
-                path,
-                fs,
-                filename,
-                partition_on,
-                write_metadata_file,
-            ],
-            toolz.merge(kwargs_pass, {"head": True}) if d == 0 else kwargs_pass,
-        )
-        part_tasks.append((name, d))
-
-    final_name = "metadata-" + name
-    # Collect metadata and write _metadata
+    data_write = df.map_partitions(
+        ToParquetFunctionWrapper(
+            engine,
+            path,
+            fs,
+            partition_on,
+            write_metadata_file,
+            i_offset,
+            name_function,
+            kwargs_pass,
+        ),
+        token="to-parquet-"
+        + tokenize(
+            df,
+            fs,
+            path,
+            append,
+            ignore_divisions,
+            partition_on,
+            division_info,
+            index_cols,
+            schema,
+        ),
+        meta=df._meta,
+        enforce_metadata=False,
+        transform_divisions=False,
+        align_dataframes=False,
+    )
 
+    # Collect metadata and write _metadata.
+    # TODO: Use tree-reduction layer (when available)
```

Comment:
Is that necessary? Doesn't seem like much of a tree, just a single reduction. Guess it's slightly more efficient?

Reply:
Why not? Right now its an all-to-one metadata reduction which we know is problematic at large scale.  If we use a formal HLG reduction layer, we will usually do the same thing, but will automatically get improved behavior at large scale.  Also, we will be removiong a "materialized layer" from the final HLG.

Reply:
Yeah, we should do that. My feeling was more that the tree-reduction layer doesn't get you many benefits at optimization compared to a low-level layer (no special fusion or anything), but I forgot that culling at least should be more efficient (when https://github.com/dask/dask/issues/8570 is resolved).